### DISCUSSION TITLE
How do I get ImGui GLFW Backend working on Android?

### DISCUSSION THREAD

**Discussion by xwellauss** [ocornut/imgui#7172] [New Users Build/Link/Run issues ONLY!]
I have managed to port a barebones version of GLFW to Android that can draw colored triangles in OpenGL and do some basic input.
https://github.com/techie-guy/glfw-android-port

I tried to use imgui on it, but it doesn't work. Maybe this is because there's some unimplemented functions in the port.

Since I'm not very familiar with the imgui codebase, Can I please get some help?

An overview of how the glfw+opengl backend works, what functions do I need to implement to get imgui working on Android?

**Comment by cos1110** [comment]
I have not used Android builds, but there is a logic behind using ImGui in any sort of library/rendering combination:
• For the ImGui library, you need to have the include paths set in place, so you can include the `imgui.h`
• Then you have the `imgui_demo` `imgui_draw` `imgui_widgets` cpp files dropped onto your project (path link not file copy) so they can be compiled from source

Then most important of all is to get the backend in place:
• I assume we talk about `imgui\backends\imgui_impl_android.h`
• You add the include path on the project
• You drop as well these files onto your project (not file copy - but file link) so they are included in the build `imgui_impl_opengl3.h` `imgui_impl_opengl3.cpp` `imgui_impl_android.h` `imgui_impl_android.cpp`

Now theoretically you are ready to go, you can use the code from here `example_android_opengl3\main.cpp` for testing purposes and hopefully all will build correctly.

**Reply by xwellauss** [reply]
I tried the Android Backend. It does work and the ImGui demo window renders correctly
![Screenshot_20231229_000644_GLFW Android Port](https://github.com/ocornut/imgui/assets/79499324/aed906bf-a2f0-4b0f-a38f-ede3c417bf45)

But I can't interact with anything(*using the touchscreen, haven't tried a mouse, Maybe that'll work*) as the Android Backend doesn't have that much features as compared to the GLFW Backend.

My GLFW port is very incomplete and I need help on what functions should I implement first to get ImGui GLFW backend to work(Rendering+Input)

**Comment by cos1110** [comment]
Good to see that you were able to setup the project sample.

If you look at the other backend implementation for reference as `imgui_impl_glfw.cpp`, you can see that the function `ImGui_ImplGlfw_NewFrame` calls these two functions as such, `ImGui_ImplGlfw_UpdateMouseData` and `ImGui_ImplGlfw_UpdateMouseCursor` that are responsible for transferring the mouse input events.

However android backend `imgui_impl_android.cpp` is based on event callbacks. I see that on `android_main` you get this callback assignment `app->onInputEvent = handleInputEvent`, that call the `ImGui_ImplAndroid_HandleInputEvent`.

Do a debug log of the event in `ImGui_ImplAndroid_HandleInputEvent` to see that the id of event is correct and then that is sent to ImGUI in the right equivalent mouse event id.

**Reply by xwellauss** [reply]
I handle the `onInputEvent` callback in GLFW (`android_init.c`).

It is not possible to call `ImGui_ImplAndroid_HandleInputEvent` from there. Even if it was possible, that wouldn't be a good solution!

So, it is not a good idea to use the Android Backend here.

I am trying to **PORT** GLFW to Android and it is at a point where it can render colored triangles and do some basic input stuff

I want to get the GLFW Backend for ImGui to work on Android. I don't know why it doesn't render anything on using the GLFW Backend(+OpenGL). I tried debugging it using RenderDoc and there were no draw calls!

Therefore, I need help here.

**Reply by xwellauss** [reply]
I managed to get the result of the above screenshot using the GLFW backend!
Thus eliminating the use of the Android backend.

The Input(using the touchscreen, haven't tried a mouse or a keyboard) still doesn't work though...

**Comment by ocornut** [comment]
Whole premise sounds bizarre. If you can port GLFW to Android you can surely improve the Android backend and that’d be much simpler.

**Reply by xwellauss** [reply]
Surely the Android Backend could be improved...

But there's no seperate backend for something like Emscripten, you just use the GLFW Backend and it just works.

Anyways, I managed to get basic input working for ImGui and I can do some clicky stuff. It "just works" using the GLFW Backend on Android!

https://github.com/ocornut/imgui/assets/79499324/9c69bd6a-79d9-4995-aa47-18a1e6df0993
